scripts/query-tracing.bcc

```diff
@@ -0,0 +1,260 @@
+  struct perf_output perf = {};
+  perf.type_ = EVENT_PARTITION_DONE;
+  perf.query_id = query_id;
+  perf.event.partition_done.partition_id = partition_id;
+  perf.event.partition_done.delta = delta;
+  perf.event.partition_done.ts = ts;
+  perf.event.partition_done.is_error = is_error;
+  events.perf_submit(ctx, &perf, sizeof(perf));    
+}
+
+void query_partition_done_handler(struct pt_regs *ctx) {
+  uint64_t query_id = 0;
+  uint64_t partition_id = 0;
+  uint64_t delta = 0;
+  uint64_t now = bpf_ktime_get_ns();
+  bpf_usdt_readarg(1, ctx, &query_id);
+  bpf_usdt_readarg(2, ctx, &partition_id);
+  bpf_usdt_readarg(3, ctx, &delta);
+  _push_partition_done(ctx, query_id, partition_id, now, delta, false);
+  return;
+}
+
+void query_partition_error_handler(struct pt_regs *ctx) {
+  uint64_t query_id = 0;
+  uint64_t partition_id = 0;
+  uint64_t delta = 0;
+  uint64_t now = bpf_ktime_get_ns();
+  bpf_usdt_readarg(1, ctx, &query_id);
+  bpf_usdt_readarg(2, ctx, &partition_id);
+  bpf_usdt_readarg(3, ctx, &delta);
+  _push_partition_done(ctx, query_id, partition_id, now, delta, true);
+  struct query_summary *summary = queries.lookup(&query_id);
+  if (!summary)
+    return;
+  summary->errors += 1;
+  return;
+}
+
+void query_done_handler(struct pt_regs *ctx) {
+  uint64_t query_id = 0;
+  bpf_usdt_readarg(1, ctx, &query_id);
+  struct query_summary *summary = queries.lookup(&query_id);
+  if (!summary)
+    return;
+  summary->done_ts = bpf_ktime_get_ns();
+  struct perf_output perf = {};
+  perf.type_ = EVENT_QUERY_STATE;
+  perf.query_id = query_id;
+  memcpy(&perf.event.query_summary, summary, sizeof(*summary));
+  queries.delete(&query_id);
+  events.perf_submit(ctx, &perf, sizeof(perf));
+}
+""",
+    usdt_contexts=[u],
+)
+
+# Unions are not magically translated, so we need to do it with ctypes manually
+
+EVENT_QUERY_STATE = 0
+EVENT_PARTITION_DONE = 1
+EVENT_INDEX_RESUME = 2
+
+
+class query_summary(ctypes.Structure):
+    _fields_ = [
+        ("start_ts", c_ulonglong),
+        ("midx_ts", c_ulonglong),
+        ("done_ts", c_ulonglong),
+        ("candidates", c_ushort),
+        ("errors", c_ushort),
+    ]
+
+
+class partition_done_event(ctypes.Structure):
+    _fields_ = [
+        ("partition_id", c_ulonglong),
+        ("ts", c_ulonglong),
+        ("delta", c_ulonglong),
+        ("is_error", c_byte),
+    ]
+
+
+class index_resume_event(ctypes.Structure):
+    _fields_ = [("ts", c_ulonglong), ("delta", c_ulonglong), ("count", c_ushort)]
+
+
+class events(ctypes.Union):
+    _fields_ = [
+        ("query_summary", query_summary),
+        ("partition_done", partition_done_event),
+        ("index_resume", index_resume_event),
+    ]
+
+
+class perf_output(ctypes.Structure):
+    _fields_ = [("query_id", c_ulonglong), ("type_", c_byte), ("event", events)]
+
+
+# Header (use R "long" format for easier subsequent plotting)
+print("variable\tvalue\tquery_id")
```

Comment:
Why didn't you also include the R script into this PR?

Reply:
There's no script currently, I did everything so far in interactive mode. I'll just add this as a comment:

```
> data = read.table('queries.txt', sep='\t', header = TRUE)
> data %>% ggplot(aes(x=value, y=variable)) + geom_density_ridges()
```